File: javascript-lib/test/ws.test.js
// Unit tests for the server-side WebSocket adapter (src/ws.js, PROTOCOL.md §10).
// Pure-function tests plus adapter tests driven over an in-memory fake socket.
import { test } from "node:test";
import assert from "node:assert/strict";
import { Readable } from "node:stream";

import {
  computeAccept,
  parseUpgrade,
  successResponse,
  encodeServerFrame,
  sniffAndUpgrade,
} from "../src/ws.js";

// A stand-in for a net.Socket: a real Readable (so read()/'readable'/'data'/
// pause/resume behave exactly as a socket's read side) that also records
// writes. Feed inbound bytes with feed().
class FakeSocket extends Readable {
  constructor() {
    super();
    this.written = [];
  }
  _read() {}
  feed(buf) {
    this.push(Buffer.from(buf));
  }
  write(buf) {
    this.written.push(Buffer.from(buf));
    return true;
  }
  setNoDelay() {
    return this;
  }
}

// Encode a masked client->server frame (all client frames must be masked).
function clientFrame(opcode, payload, { fin = true, mask = Buffer.from([0xa1, 0xb2, 0xc3, 0xd4]) } = {}) {
  const n = payload.length;
  let header;
  const b0 = (fin ? 0x80 : 0) | opcode;
  if (n < 126) header = Buffer.from([b0, 0x80 | n]);
  else if (n < 65536) header = Buffer.from([b0, 0x80 | 126, (n >> 8) & 0xff, n & 0xff]);
  else {
    header = Buffer.alloc(10);
    header[0] = b0;
    header[1] = 0x80 | 127;
    header.writeBigUInt64BE(BigInt(n), 2);
  }
  const masked = Buffer.allocUnsafe(n);
  for (let i = 0; i < n; i++) masked[i] = payload[i] ^ mask[i & 3];
  return Buffer.concat([header, mask, masked]);
}

const VALID_REQUEST = Buffer.from(
  "GET /chat HTTP/1.1\r\nHost: x\r\nUpgrade: websocket\r\nConnection: keep-alive, Upgrade\r\n" +
    "Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\nSec-WebSocket-Version: 13\r\n\r\n",
  "latin1",
);

test("accept key matches the RFC 6455 vector", () => {
  assert.equal(computeAccept("dGhlIHNhbXBsZSBub25jZQ=="), "s3pPLMBiTxaQ9kYGzzhZRbK+xOo=");
});

test("parseUpgrade returns the key for a valid request", () => {
  assert.equal(parseUpgrade(VALID_REQUEST), "dGhlIHNhbXBsZSBub25jZQ==");
});

test("parseUpgrade rejects bad requests", () => {
  const bad = [
    ["POST / HTTP/1.1\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\nSec-WebSocket-Version: 13\r\n\r\n", "400"],
    ["GET / HTTP/1.1\r\nConnection: Upgrade\r\nSec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\nSec-WebSocket-Version: 13\r\n\r\n", "400"],
    ["GET / HTTP/1.1\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\nSec-WebSocket-Version: 8\r\n\r\n", "426"],
    ["GET / HTTP/1.1\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Key: short\r\nSec-WebSocket-Version: 13\r\n\r\n", "400"],
  ];
  for (const [req, status] of bad) {
    assert.throws(() => parseUpgrade(req), (e) => e.message.startsWith(status));
  }
});

// Collect a fixed number of stream bytes from a WsDuplex.
function readN(adapter, n) {
  return new Promise((resolve) => {
    const chunks = [];
    let total = 0;
    adapter.on("data", (c) => {
      chunks.push(c);
      total += c.length;
      if (total >= n) resolve(Buffer.concat(chunks).subarray(0, n));
    });
  });
}

async function wsAdapterOver(frames) {
  const raw = new FakeSocket();
  const p = sniffAndUpgrade(raw);
  raw.feed(Buffer.concat([VALID_REQUEST, frames]));
  const result = await p;
  return { adapter: result.socket, raw };
}

test("sniff returns the sniffed bytes as prebuffer on the raw path", async () => {
  const raw = new FakeSocket();
  const p = sniffAndUpgrade(raw);
  raw.feed(Buffer.from([0x00, 0x00, 0x00, 0x12, 0x99]));
  const result = await p;
  assert.equal(result.kind, "raw");
  assert.equal(result.socket, raw);
  assert.deepEqual(result.prebuffer, Buffer.from([0x00, 0x00, 0x00, 0x12, 0x99]));
});

test("sniff rejects an unknown first byte", async () => {
  const raw = new FakeSocket();
  const p = sniffAndUpgrade(raw);
  raw.feed(Buffer.from([0x99, 1, 2]));
  assert.equal(await p, null);
  assert.ok(raw.destroyed);
});

test("upgrade responds 101 and streams binary payloads", async () => {
  const { adapter, raw } = await wsAdapterOver(clientFrame(0x2, Buffer.from("hello")));
  const resp = Buffer.concat(raw.written).toString("latin1");
  assert.ok(resp.startsWith("HTTP/1.1 101 Switching Protocols\r\n"));
  assert.ok(resp.includes("Sec-WebSocket-Accept: s3pPLMBiTxaQ9kYGzzhZRbK+xOo=\r\n"));
  assert.deepEqual(await readN(adapter, 5), Buffer.from("hello"));
});

test("bad upgrade returns null and writes 400", async () => {
  const raw = new FakeSocket();
  const p = sniffAndUpgrade(raw);
  raw.feed(Buffer.from("GET / HTTP/1.1\r\nUpgrade: websocket\r\n\r\n", "latin1"));
  assert.equal(await p, null);
  assert.ok(Buffer.concat(raw.written).toString("latin1").startsWith("HTTP/1.1 400 Bad Request\r\n"));
});

test("masked binary is unmasked into the stream", async () => {
  const { adapter } = await wsAdapterOver(clientFrame(0x2, Buffer.from("aiomsg-payload")));
  assert.deepEqual(await readN(adapter, 14), Buffer.from("aiomsg-payload"));
});

test("fragmented binary reassembles across boundaries", async () => {
  const frames = Buffer.concat([
    clientFrame(0x2, Buffer.from("abc"), { fin: false }),
    clientFrame(0x0, Buffer.from("def"), { fin: false }),
    clientFrame(0x0, Buffer.from("ghi"), { fin: true }),
  ]);
  const { adapter } = await wsAdapterOver(frames);
  assert.deepEqual(await readN(adapter, 9), Buffer.from("abcdefghi"));
});

test("ping is answered with a pong then data flows", async () => {
  const frames = Buffer.concat([clientFrame(0x9, Buffer.from("pingdata")), clientFrame(0x2, Buffer.from("XY"))]);
  const { adapter, raw } = await wsAdapterOver(frames);
  assert.deepEqual(await readN(adapter, 2), Buffer.from("XY"));
  const pong = encodeServerFrame(0xa, Buffer.from("pingdata"));
  assert.ok(raw.written.some((w) => w.equals(pong)));
});

test("close frame is echoed and ends the stream", async () => {
  const { adapter, raw } = await wsAdapterOver(clientFrame(0x8, Buffer.from([0x03, 0xe8])));
  await new Promise((res) => adapter.once("close", res));
  const closeFrame = encodeServerFrame(0x8, Buffer.from([0x03, 0xe8]));
  assert.ok(raw.written.some((w) => w.equals(closeFrame)));
});

test("text frame is rejected with 1003", async () => {
  const { adapter, raw } = await wsAdapterOver(clientFrame(0x1, Buffer.from("hi")));
  await new Promise((res) => adapter.once("close", res));
  const closeFrame = encodeServerFrame(0x8, Buffer.from([0x03, 0xeb])); // 1003
  assert.ok(raw.written.some((w) => w.equals(closeFrame)));
});

test("unmasked client frame is rejected with 1002", async () => {
  const raw = new FakeSocket();
  const p = sniffAndUpgrade(raw);
  // Unmasked binary frame (MASK bit clear) — illegal from a client.
  const bad = Buffer.concat([Buffer.from([0x82, 0x04]), Buffer.from("nope")]);
  raw.feed(Buffer.concat([VALID_REQUEST, bad]));
  const { socket: adapter } = await p;
  await new Promise((res) => adapter.once("close", res));
  const closeFrame = encodeServerFrame(0x8, Buffer.from([0x03, 0xea])); // 1002
  assert.ok(raw.written.some((w) => w.equals(closeFrame)));
});

test("write emits one binary frame per aiomsg write", async () => {
  const { adapter, raw } = await wsAdapterOver(Buffer.alloc(0));
  raw.written.length = 0; // drop the 101 response
  adapter.write(Buffer.from("\x00\x00\x00\x03abc", "latin1"));
  await new Promise((res) => setImmediate(res));
  assert.deepEqual(raw.written[0], encodeServerFrame(0x2, Buffer.from("\x00\x00\x00\x03abc", "latin1")));
});

File: java-lib/src/test/java/aiomsg/WsTest.java
// Unit tests for the server-side WebSocket adapter (Ws, PROTOCOL.md §10).
// Pure-function tests plus adapter tests over in-memory streams — no sockets.
package aiomsg;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import org.junit.jupiter.api.Test;

class WsTest {

    private static final byte[] MASK = {(byte) 0xa1, (byte) 0xb2, (byte) 0xc3, (byte) 0xd4};

    // Encode a masked client->server frame (all client frames must be masked).
    private static byte[] clientFrame(int opcode, byte[] payload, boolean fin) {
        int n = payload.length;
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        out.write((fin ? 0x80 : 0x00) | opcode);
        if (n < 126) {
            out.write(0x80 | n);
        } else if (n < 65536) {
            out.write(0x80 | 126);
            out.write((n >>> 8) & 0xff);
            out.write(n & 0xff);
        } else {
            out.write(0x80 | 127);
            for (int i = 0; i < 8; i++) out.write((int) (((long) n) >>> (8 * (7 - i))) & 0xff);
        }
        out.writeBytes(MASK);
        for (int i = 0; i < n; i++) out.write((payload[i] ^ MASK[i & 3]) & 0xff);
        return out.toByteArray();
    }

    private static byte[] clientFrame(int opcode, byte[] payload) {
        return clientFrame(opcode, payload, true);
    }

    // Read up to n bytes from an adapter over the given inbound frames; also
    // return the bytes written back (pongs, close frames).
    private record Result(byte[] read, byte[] written) {}

    private static Result driveRead(byte[] frames, int n) throws IOException {
        ByteArrayOutputStream written = new ByteArrayOutputStream();
        Ws.Framed framed = new Ws.Framed(new ByteArrayInputStream(frames), written);
        InputStream in = framed.inputStream();
        ByteArrayOutputStream got = new ByteArrayOutputStream();
        int b;
        while (got.size() < n && (b = in.read()) >= 0) got.write(b);
        return new Result(got.toByteArray(), written.toByteArray());
    }

    // Drain the adapter to EOF, returning what was written back.
    private static byte[] driveToEof(byte[] frames) throws IOException {
        ByteArrayOutputStream written = new ByteArrayOutputStream();
        Ws.Framed framed = new Ws.Framed(new ByteArrayInputStream(frames), written);
        InputStream in = framed.inputStream();
        while (in.read() >= 0) { /* discard */ }
        return written.toByteArray();
    }

    @Test
    void acceptKeyMatchesRfcVector() {
        assertEquals("s3pPLMBiTxaQ9kYGzzhZRbK+xOo=", Ws.computeAccept("dGhlIHNhbXBsZSBub25jZQ=="));
    }

    @Test
    void parseUpgradeReturnsKey() throws Exception {
        byte[] req = ("GET /chat HTTP/1.1\r\nHost: x\r\nUpgrade: websocket\r\n"
                + "Connection: keep-alive, Upgrade\r\nSec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\n"
                + "Sec-WebSocket-Version: 13\r\nOrigin: http://evil\r\n\r\n").getBytes(StandardCharsets.ISO_8859_1);
        assertEquals("dGhlIHNhbXBsZSBub25jZQ==", Ws.parseUpgrade(req));
    }

    @Test
    void parseUpgradeRejectsBadRequests() {
        String[][] bad = {
            {"POST / HTTP/1.1\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\nSec-WebSocket-Version: 13\r\n\r\n", "400"},
            {"GET / HTTP/1.1\r\nConnection: Upgrade\r\nSec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\nSec-WebSocket-Version: 13\r\n\r\n", "400"},
            {"GET / HTTP/1.1\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\nSec-WebSocket-Version: 8\r\n\r\n", "426"},
            {"GET / HTTP/1.1\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Key: short\r\nSec-WebSocket-Version: 13\r\n\r\n", "400"},
        };
        for (String[] c : bad) {
            Exception e = assertThrows(Exception.class,
                    () -> Ws.parseUpgrade(c[0].getBytes(StandardCharsets.ISO_8859_1)));
            assertTrue(e.getMessage().startsWith(c[1]), e.getMessage());
        }
    }

    @Test
    void maskedBinaryIsUnmasked() throws IOException {
        byte[] payload = "aiomsg-payload".getBytes(StandardCharsets.UTF_8);
        Result r = driveRead(clientFrame(0x2, payload), payload.length);
        assertArrayEquals(payload, r.read());
    }

    @Test
    void fragmentedBinaryReassembles() throws IOException {
        ByteArrayOutputStream frames = new ByteArrayOutputStream();
        frames.writeBytes(clientFrame(0x2, "abc".getBytes(), false));
        frames.writeBytes(clientFrame(0x0, "def".getBytes(), false));
        frames.writeBytes(clientFrame(0x0, "ghi".getBytes(), true));
        Result r = driveRead(frames.toByteArray(), 9);
        assertArrayEquals("abcdefghi".getBytes(), r.read());
    }

    @Test
    void pingIsAnsweredWithPong() throws IOException {
        ByteArrayOutputStream frames = new ByteArrayOutputStream();
        frames.writeBytes(clientFrame(0x9, "pingdata".getBytes()));
        frames.writeBytes(clientFrame(0x2, "XY".getBytes()));
        Result r = driveRead(frames.toByteArray(), 2);
        assertArrayEquals("XY".getBytes(), r.read());
        assertArrayEquals(Ws.serverFrame(0xA, "pingdata".getBytes()), r.written());
    }

    @Test
    void closeFrameIsEchoedAndEndsStream() throws IOException {
        byte[] written = driveToEof(clientFrame(0x8, new byte[]{0x03, (byte) 0xe8})); // 1000
        assertArrayEquals(Ws.serverFrame(0x8, new byte[]{0x03, (byte) 0xe8}), written);
    }

    @Test
    void textFrameRejectedWith1003() throws IOException {
        byte[] written = driveToEof(clientFrame(0x1, "hi".getBytes()));
        assertArrayEquals(Ws.serverFrame(0x8, new byte[]{0x03, (byte) 0xeb}), written); // 1003
    }

    @Test
    void unmaskedFrameRejectedWith1002() throws IOException {
        // Unmasked binary frame (MASK bit clear) — illegal from a client.
        byte[] bad = new byte[]{(byte) 0x82, 0x04, 'n', 'o', 'p', 'e'};
        byte[] written = driveToEof(bad);
        assertArrayEquals(Ws.serverFrame(0x8, new byte[]{0x03, (byte) 0xea}), written); // 1002
    }

    @Test
    void oneBinaryFramePerAiomsgWrite() throws IOException {
        ByteArrayOutputStream written = new ByteArrayOutputStream();
        Ws.Framed framed = new Ws.Framed(new ByteArrayInputStream(new byte[0]), written);
        byte[] msg = {0, 0, 0, 3, 'a', 'b', 'c'};
        framed.outputStream().write(msg);
        assertArrayEquals(Ws.serverFrame(0x2, msg), written.toByteArray());
    }
}
